File: src/main/java/aurocosh/divinefavor/common/potions/potions/PotionGroundFlow.java
package aurocosh.divinefavor.common.potions.potions;

import aurocosh.divinefavor.common.config.common.ConfigSpells;
import aurocosh.divinefavor.common.network.message.client.MessageSyncFlyingCapability;
import aurocosh.divinefavor.common.potions.base.potion.ModPotionToggle;
import aurocosh.divinefavor.common.potions.common.ModPotions;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.math.BlockPos;

public class PotionGroundFlow extends ModPotionToggle {
    public PotionGroundFlow() {
        super("ground_flow", true, 0x7FB8A4);
    }

    @Override
    public void performEffect(EntityLivingBase livingBase, int amplifier) {
        assert !livingBase.world.isRemote;

        BlockPos pos = livingBase.getPosition();
        EntityPlayer player = (EntityPlayer) livingBase;
        if(!player.capabilities.allowFlying && pos.getY() <= ConfigSpells.groundFlow.yLimit)
            player.capabilities.allowFlying = true;
        else if(player.capabilities.allowFlying && pos.getY() > ConfigSpells.groundFlow.yLimit) {
            player.capabilities.allowFlying = false;
            player.capabilities.isFlying = false;
            player.removePotionEffect(ModPotions.ground_flow);
            new MessageSyncFlyingCapability(false).sendTo(player);
        }
    }

    @Override
    public boolean isReady(int duration, int amplifier) {
        return true;
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/potions/potions/PotionGroundFlow.kt
package aurocosh.divinefavor.common.potions.potions

import aurocosh.divinefavor.common.config.common.ConfigSpell
import aurocosh.divinefavor.common.potions.base.potion.ModPotionToggle
import aurocosh.divinefavor.common.potions.common.ModPotions
import aurocosh.divinefavor.common.util.UtilPlayer
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer

class PotionGroundFlow : ModPotionToggle("ground_flow", 0x7FB8A4) {

    override fun onPotionAdded(livingBase: EntityLivingBase) {
        super.onPotionAdded(livingBase)
        if (livingBase !is EntityPlayer)
            livingBase.removePotionEffect(ModPotions.ground_flow)
    }

    override fun onPotionRemoved(livingBase: EntityLivingBase) {
        super.onPotionRemoved(livingBase)
        if (livingBase is EntityPlayer)
            UtilPlayer.setAllowFlying(livingBase, false)
    }

    override fun performEffect(livingBase: EntityLivingBase, amplifier: Int) {
        if (livingBase.world.isRemote)
            return
        if (livingBase !is EntityPlayer)
            return
        val allowFlying = livingBase.position.y <= ConfigSpell.groundFlow.yLimit
        UtilPlayer.setAllowFlying(livingBase, allowFlying)
        if (!allowFlying)
            livingBase.removePotionEffect(ModPotions.ground_flow)
    }

    override fun isReady(duration: Int, amplifier: Int): Boolean {
        return true
    }
}
